File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    Map.new(remove_special_chars(sentence), fn x ->
      { x, occurences_in(x, remove_special_chars(sentence)) }
    end)
  end

  defp occurences_in(needle, haystack) do
    Enum.count(haystack, fn(item) -> item == needle end)
  end

  defp remove_special_chars(sentence) do
    String.replace(sentence, ~r/[!_,&@$%^:\s]/, " ") |> String.downcase |> String.split
  end
end

File: ruby/word-count/word_count.rb
class Phrase
  def initialize(sentence)
    @words = sentence.split(/[^'|\w]/).reject { |w| w.empty? }.map { |w| w.downcase.strip("'") }
  end

  def word_count
    @words.uniq.map { |w| [w, @words.count(w)] }.to_h
  end
end
